Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,398 @@
+
+            // Queue the adapt operation and trigger refresh operation if credential is requires refresh
+            guard !credential.requiresRefresh else {
+                let operation = AdaptOperation(urlRequest: urlRequest, session: session, completion: completion)
+                mutableState.adaptOperations.append(operation)
+                refresh(credential, for: session, insideLock: &mutableState)
+                return .adaptDeferred
+            }
+
+            return .adapt(credential)
+        }
+
+        switch adaptResult {
+        case .adapt(let credential):
+            var authenticatedRequest = urlRequest
+            authenticator.apply(credential, to: &authenticatedRequest)
+            completion(.success(authenticatedRequest))
+
+        case .doNotAdapt(let adaptError):
+            completion(.failure(adaptError))
+
+        case .adaptDeferred:
+            // No-op: adapt operation captured during refresh
+            break
+        }
+    }
+
+    // MARK: Retry
+
+    public func retry(_ request: Request, for session: Session, dueTo error: Error, completion: @escaping (RetryResult) -> Void) {
+        // Do not attempt retry if there was not an original request and response from the server
+        guard let urlRequest = request.request, let response = request.response else {
+            completion(.doNotRetry)
+            return
+        }
+
+        // Do not attempt retry unless inspector verifies failure was due to authentication error (i.e. 401 status code)
+        guard authenticator.didRequest(urlRequest, with: response, failDueToAuthenticationError: error) else {
+            completion(.doNotRetry)
+            return
+        }
+
+        // Do not attempt retry if there is no credential (throw missing credential error instead since retry required)
+        guard let credential = credential else {
+            let error = AuthenticationError.missingCredential
+            completion(.doNotRetryWithError(error))
+            return
+        }
+
+        // Retry the request if inspector verifies it was authenticated with a previous credential
+        guard authenticator.isRequest(urlRequest, authenticatedWith: credential) else {
+            completion(.retry)
+            return
+        }
+
+        $mutableState.write { mutableState in
+            mutableState.requestsToRetry.append(completion)
+
+            guard !mutableState.isRefreshing else { return }
+
+            refresh(credential, for: session, insideLock: &mutableState)
+        }
+    }
+
+    // MARK: Refresh
+
+    private func refresh(_ credential: Credential, for session: Session, insideLock mutableState: inout MutableState) {
+        guard !isRefreshExcessive(insideLock: &mutableState) else {
+            let error = AuthenticationError.excessiveRefresh
+            handleRefreshFailure(error, insideLock: &mutableState)
+            return
+        }
+
+        mutableState.refreshTimestamps.append(Date())
```

Comment:
I don't think this is necessary. `Date` is a point in UTC time that's consistent globally that does not account for calendar or time zone. So I don't think DST is a concern from what I can tell. Do you have some documentation somewhere that points out this behavior? Everything in the `Date` documentation leads me to believe this is not the case.

Reply:
I think we still want to use a monotonic source here, as `Date` is subject to automatic and manual updates on device which would otherwise throw things off here. Minor, but could be noticeable if the user's device has a lot of clock skew, or they update their date manually.

Reply:
Gotcha. Will update to use uptime for the internal timestamps.

Reply:
Updated to `systemUptime` in b9ce5df4.